### PR TITLE
[CORE-1116] S3 gateway and projects

### PR DESCRIPTION
**Pull request by robert-uhl**

### CODE REVIEW

File: src/server/pfs/s3/driver.go [hunk 1]
```diff
@@ -52,13 +53,31 @@ func NewMasterDriver() *MasterDriver {
...
+
 func (d *MasterDriver) listBuckets(pc *client.APIClient, r *http.Request, buckets *[]*s2.Bucket) error {
+	var isProjectAware = mux.Vars(r)["isProjectAware"] == "yes"
 	repos, err := pc.ListRepoByType("") // get repos of all types
 	if err != nil {
 		return err
 	}
 
 	for _, repo := range repos {
+		if !isProjectAware && repo.Repo.Project.Name != pfs.DefaultProjectName {
```
Comment (bbonenfant) [inline src/server/pfs/s3/driver.go:78]: Should this be `isProjectAware` rather than `!isProjectAware`?

**Review by bbonenfant — APPROVED** [review]
This looks OK to me. Just to clarify, this is a breaking change and the user must always specify the project even if its the default project?

**Comment by robert-uhl** [comment]
> This looks OK to me. Just to clarify, this is a breaking change and the user must always specify the project even if its the default project?

Nope, it is backwards-compatible. If the user sets the access key ID and secret access key to the same value (as is current), then the existing non-project-aware code is in effect; if the user prepends `"PAC1"` to the access key ID, then the new project-aware code is in effect.

This way users can opt in each S3 gateway client one at a time.

File: src/server/pfs/s3/driver.go (hunk 1 above)
Comment (robert-uhl) [reply src/server/pfs/s3/driver.go:78]: No: the logic here is that if the client is not project aware and if the repo is not in the default project, then the repo is skipped. So only repos in the default project will be returned to non-project-aware clients.

**Pull request merged**